File: backend-java/src/main/java/mx/codemx/modules/duelos/ClaudeGeneradorProblemas.java
package mx.codemx.modules.duelos;

import com.anthropic.client.AnthropicClient;
import com.anthropic.client.okhttp.AnthropicOkHttpClient;
import com.anthropic.models.messages.Message;
import com.anthropic.models.messages.MessageCreateParams;
import com.anthropic.models.messages.TextBlock;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;
import java.util.stream.Collectors;
import mx.codemx.modules.retos.CasoPrueba;
import mx.codemx.modules.retos.Dificultad;
import org.springframework.stereotype.Component;
import tools.jackson.databind.ObjectMapper;

@Component
public class ClaudeGeneradorProblemas implements GeneradorProblemas {

    private static final String MODELO = "claude-opus-4-8";

    private static final List<String> TEMAS = List.of(
            "cadenas de texto", "matemática básica", "arreglos y listas", "lógica condicional",
            "ciclos y conteo", "ordenamiento simple", "búsqueda", "geometría sencilla",
            "manejo de dígitos", "secuencias numéricas");

    private final ObjectMapper json = new ObjectMapper();

    @Override
    public ProblemaDuelo generar(Dificultad dificultad) {
        String tema = TEMAS.get(ThreadLocalRandom.current().nextInt(TEMAS.size()));
        Estilo estilo = estiloDe(dificultad);

        String instruccion = """
                Eres el juez de una competencia de programación 1 vs 1 en español. Genera UN problema de \
                dificultad %s sobre el tema "%s". %s

                Devuelve EXCLUSIVAMENTE un objeto JSON válido (sin texto adicional, sin ```), con esta forma:
                {
                  "titulo": string corto,
                  "enunciado": string claro en español que describa qué imprimir (y la entrada, si la hay),
                  "ejemploEntrada": string (vacío si el problema no lee nada),
                  "ejemploSalida": string,
                  "casos": [ { "entrada": string, "salidaEsperada": string }, ... ]
                }

                Reglas estrictas:
                %s
                - Cada "salidaEsperada" debe ser EXACTAMENTE lo que un programa correcto imprime para esa \
                entrada, sin espacios ni saltos de línea finales sobrantes.
                - El problema debe poder resolverse en menos de 2 segundos y sin librerías externas.
                - No incluyas la solución ni explicaciones, sólo el JSON.\
                """.formatted(estilo.nivel(), tema, estilo.formato(), estilo.reglaCasos());

        AnthropicClient client = AnthropicOkHttpClient.fromEnv();

        MessageCreateParams params = MessageCreateParams.builder()
                .model(MODELO)
                .maxTokens(2048)
                .addUserMessage(instruccion)
                .build();

        Message respuesta = client.messages().create(params);

        String texto = respuesta.content().stream()
                .map(bloque -> bloque.text().map(TextBlock::text).orElse(""))
                .collect(Collectors.joining());

        return parsear(texto);
    }

    private static Estilo estiloDe(Dificultad dificultad) {
        return switch (dificultad) {
            case BASICO -> new Estilo(
                    "muy fácil, para alguien que apenas empieza en Python",
                    "El alumno SOLO debe usar variables y print(). NO debe leer nada de la entrada estándar "
                            + "(nada de input()): la salida es fija. El problema es de 1 o 2 líneas (p. ej. crear una "
                            + "variable e imprimirla, o imprimir la suma de dos números dados en el enunciado).",
                    "- Incluye EXACTAMENTE 1 caso con \"entrada\" vacía (\"\") y la \"salidaEsperada\" fija.");
            case AVANZADO -> new Estilo(
                    "difícil (requiere pensar el algoritmo: ciclos, condiciones, etc.)",
                    "Se resuelve en Python leyendo de la entrada estándar (stdin) e imprimiendo en stdout.",
                    "- Incluye entre 5 y 7 casos de prueba deterministas que cubran casos límite.");
            case INTERMEDIO -> new Estilo(
                    "intermedio (una entrada simple con input() y quizá una condición)",
                    "Se resuelve en Python leyendo de la entrada estándar (stdin) e imprimiendo en stdout.",
                    "- Incluye entre 4 y 6 casos de prueba deterministas que cubran casos límite.");
        };
    }

    private ProblemaDuelo parsear(String texto) {
        int inicio = texto.indexOf('{');
        int fin = texto.lastIndexOf('}');
        if (inicio < 0 || fin <= inicio) {
            throw new IllegalStateException("Claude no devolvió un JSON de problema válido.");
        }

        ProblemaDto dto;
        try {
            dto = json.readValue(texto.substring(inicio, fin + 1), ProblemaDto.class);
        } catch (Exception e) {
            throw new IllegalStateException("No se pudo deserializar el problema de Claude.", e);
        }

        List<CasoPrueba> casos = new ArrayList<>();
        if (dto.casos != null) {
            for (CasoDto c : dto.casos) {
                casos.add(new CasoPrueba(
                        c.entrada == null ? "" : c.entrada,
                        c.salidaEsperada == null ? "" : c.salidaEsperada.stripTrailing()));
            }
        }

        if (casos.isEmpty()) {
            throw new IllegalStateException("El problema de Claude no trae casos de prueba.");
        }

        return new ProblemaDuelo(
                dto.titulo == null ? "Reto relámpago" : dto.titulo,
                dto.enunciado == null ? "" : dto.enunciado,
                dto.ejemploEntrada == null ? casos.get(0).getInputData() : dto.ejemploEntrada,
                dto.ejemploSalida == null ? casos.get(0).getOutputEsperado() : dto.ejemploSalida,
                casos);
    }

    private record Estilo(String nivel, String formato, String reglaCasos) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private static final class ProblemaDto {
        public String titulo;
        public String enunciado;
        public String ejemploEntrada;
        public String ejemploSalida;
        public List<CasoDto> casos;
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private static final class CasoDto {
        public String entrada;
        public String salidaEsperada;
    }
}

File: backend/Modules/Duelos/ClaudeGeneradorProblemas.cs
using System.Text.Json;
using Anthropic;
using Anthropic.Models.Messages;
using CodeMX.Api.Modules.Retos;

namespace CodeMX.Api.Modules.Duelos;

/// <summary>
/// Genera el problema del duelo con Claude (SDK oficial de C#, modelo claude-opus-4-8).
/// Claude inventa el reto, su solución de referencia y los casos de prueba; la corrección
/// se decide después ejecutando el código de cada jugador contra esos casos (motor Python),
/// así que la revisión es determinista y no depende de una nota subjetiva del modelo.
///
/// Requiere la variable de entorno ANTHROPIC_API_KEY. Si no está o la llamada falla,
/// <see cref="GeneradorProblemas"/> recurre automáticamente al respaldo.
/// </summary>
public class ClaudeGeneradorProblemas : IGeneradorProblemas
{
    private const string Modelo = "claude-opus-4-8";

    // Temas para variar los duelos en cada partida.
    private static readonly string[] Temas =
    {
        "cadenas de texto", "matemática básica", "arreglos y listas", "lógica condicional",
        "ciclos y conteo", "ordenamiento simple", "búsqueda", "geometría sencilla",
        "manejo de dígitos", "secuencias numéricas"
    };

    public async Task<ProblemaDuelo> GenerarAsync(Dificultad dificultad, CancellationToken ct = default)
    {
        var tema = Temas[Random.Shared.Next(Temas.Length)];

        // El nivel cambia el ESTILO del problema, no sólo el adjetivo: el "fácil" para un
        // principiante no usa entrada estándar (solo print y variables); eso se aprende después.
        var (nivel, formato, reglaCasos) = dificultad switch
        {
            Dificultad.BASICO => (
                "muy fácil, para alguien que apenas empieza en Python",
                "El alumno SOLO debe usar variables y print(). NO debe leer nada de la entrada estándar " +
                "(nada de input()): la salida es fija. El problema es de 1 o 2 líneas (p. ej. crear una " +
                "variable e imprimirla, o imprimir la suma de dos números dados en el enunciado).",
                "- Incluye EXACTAMENTE 1 caso con \"entrada\" vacía (\"\") y la \"salidaEsperada\" fija."),
            Dificultad.AVANZADO => (
                "difícil (requiere pensar el algoritmo: ciclos, condiciones, etc.)",
                "Se resuelve en Python leyendo de la entrada estándar (stdin) e imprimiendo en stdout.",
                "- Incluye entre 5 y 7 casos de prueba deterministas que cubran casos límite."),
            _ => (
                "intermedio (una entrada simple con input() y quizá una condición)",
                "Se resuelve en Python leyendo de la entrada estándar (stdin) e imprimiendo en stdout.",
                "- Incluye entre 4 y 6 casos de prueba deterministas que cubran casos límite.")
        };

        // El cliente lee ANTHROPIC_API_KEY del entorno automáticamente.
        var client = new AnthropicClient();

        var instruccion =
            $"Eres el juez de una competencia de programación 1 vs 1 en español. Genera UN problema de " +
            $"dificultad {nivel} sobre el tema \"{tema}\". {formato}\n\n" +
            "Devuelve EXCLUSIVAMENTE un objeto JSON válido (sin texto adicional, sin ```), con esta forma:\n" +
            "{\n" +
            "  \"titulo\": string corto,\n" +
            "  \"enunciado\": string claro en español que describa qué imprimir (y la entrada, si la hay),\n" +
            "  \"ejemploEntrada\": string (vacío si el problema no lee nada),\n" +
            "  \"ejemploSalida\": string,\n" +
            "  \"casos\": [ { \"entrada\": string, \"salidaEsperada\": string }, ... ]\n" +
            "}\n\n" +
            "Reglas estrictas:\n" +
            reglaCasos + "\n" +
            "- Cada \"salidaEsperada\" debe ser EXACTAMENTE lo que un programa correcto imprime para esa entrada, " +
            "sin espacios ni saltos de línea finales sobrantes.\n" +
            "- El problema debe poder resolverse en menos de 2 segundos y sin librerías externas.\n" +
            "- No incluyas la solución ni explicaciones, sólo el JSON.";

        MessageCreateParams parameters = new()
        {
            MaxTokens = 2048,
            Model = Modelo,
            Messages = [new() { Role = Role.User, Content = instruccion }],
        };

        var respuesta = await client.Messages.Create(parameters, cancellationToken: ct);

        var texto = string.Join(
            "",
            respuesta.Content.Select(b => b.Value).OfType<TextBlock>().Select(t => t.Text));

        return Parsear(texto);
    }

    /// <summary>Extrae el objeto JSON de la respuesta y lo convierte en un ProblemaDuelo.</summary>
    private static ProblemaDuelo Parsear(string texto)
    {
        int inicio = texto.IndexOf('{');
        int fin = texto.LastIndexOf('}');
        if (inicio < 0 || fin <= inicio)
            throw new InvalidOperationException("Claude no devolvió un JSON de problema válido.");

        var json = texto.Substring(inicio, fin - inicio + 1);
        var dto = JsonSerializer.Deserialize<ProblemaDto>(json,
            new JsonSerializerOptions { PropertyNameCaseInsensitive = true })
            ?? throw new InvalidOperationException("No se pudo deserializar el problema de Claude.");

        var casos = (dto.Casos ?? new())
            .Select(c => new CasoPrueba { InputData = c.Entrada ?? "", OutputEsperado = (c.SalidaEsperada ?? "").TrimEnd() })
            .ToList();

        if (casos.Count == 0)
            throw new InvalidOperationException("El problema de Claude no trae casos de prueba.");

        return new ProblemaDuelo(
            dto.Titulo ?? "Reto relámpago",
            dto.Enunciado ?? "",
            dto.EjemploEntrada ?? casos[0].InputData ?? "",
            dto.EjemploSalida ?? casos[0].OutputEsperado,
            casos);
    }

    // DTOs sólo para deserializar la respuesta de Claude.
    private sealed class ProblemaDto
    {
        public string? Titulo { get; set; }
        public string? Enunciado { get; set; }
        public string? EjemploEntrada { get; set; }
        public string? EjemploSalida { get; set; }
        public List<CasoDto>? Casos { get; set; }
    }

    private sealed class CasoDto
    {
        public string? Entrada { get; set; }
        public string? SalidaEsperada { get; set; }
    }
}
